CoreWordCloud/src/au/gov/asd/tac/constellation/views/wordcloud/content/ContentPairwiseSimilarityServices.java

```diff
@@ -0,0 +1,334 @@
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ *
+ * @author twilight_sparkle
+ * @author Delphinus8821
+ */
+public class ContentPairwiseSimilarityServices {
+    
+    private static final String UNSUPPORTED = "This calculator does not compute modulii";
+
+    public static class MutableDouble {
+
+        private double val;
+
+        public MutableDouble(final double val) {
+            this.val = val;
+        }
+    }
+
+    private PairwiseSimilarityCalculator similarityCalculator;
+    private final PairwiseComparisonTokenHandler handler;
+    private final List<ElementSimilarity> pairwiseSimilarities = Collections.synchronizedList(new LinkedList<>());
+    private final Map<Integer, Double> modulii = new HashMap<>();
+
+    private ContentPairwiseSimilarityServices(final PairwiseComparisonTokenHandler handler) {
+        this.handler = handler;
+        pairwiseSimilaritiesCalculation = new HashMap[handler.elementCapacity];
+    }
+
+    public static Map<Integer, Integer> clusterSimilarElements(final PairwiseComparisonTokenHandler handler, final NGramAnalysisParameters nGramParams) {
+        return pairwiseComparison(handler, nGramParams).cluster();
+    }
+
+    public static List<ElementSimilarity> scoreSimilarPairs(final PairwiseComparisonTokenHandler handler, final NGramAnalysisParameters nGramParams) {
+        return pairwiseComparison(handler, nGramParams).pairwiseSimilarities;
+    }
+
+    private static ContentPairwiseSimilarityServices pairwiseComparison(final PairwiseComparisonTokenHandler handler, final NGramAnalysisParameters nGramParams) {
+        final ContentPairwiseSimilarityServices cpss = new ContentPairwiseSimilarityServices(handler);
+        if (nGramParams.isBinarySpace()) {
+            cpss.similarityCalculator = new BinarySpaceTaxicabNormCalculator();
+        } else {
+            cpss.similarityCalculator = new IntegerSpaceTaxicabNormCalculator();
+        }
+        cpss.computeSimilarities(nGramParams.getThreshold());
+        return cpss;
+    }
+
+    private void computeSimilarities(final double threshold) {
+
+        if (similarityCalculator.computeModulii()) {
+            for (int i = 0; i < handler.totalChunks; i++) {
+                final int currentChunk = i;
+                handler.tokenElementFrequencies.values().parallelStream().forEach(value -> computeModulii(value[currentChunk]));
+            }
+        }
+
+        for (int i = 0; i < handler.numChunks; i++) {
+            int j;
+            final int max;
+            if (handler.elementsOfInterest == null) {
+                j = i;
+                max = handler.numChunks;
+            } else {
+                j = handler.numChunks;
+                max = handler.totalChunks;
+            }
+            for (; j < max; j++) {
```

Comment:
seems like this should be a while loop (or a different loop variable should be used which is initialised to j